### PR TITLE
Save docker-compose logs from integration tests.

### PR DESCRIPTION
**Pull request by cmacknz**
## What does this PR do?

Captures and preserves the docker-compose logs after running integration or system tests. Allows auditing service logs for issues, in particular it enables auditing the Elasticsearch logs for deprecation warnings after test runs.

## Why is it important?

We need to ensure we aren't using any soon to be deprecated Elasticsearch APIs.

## Checklist

- [ x ] My code follows the style guidelines of this project
- [ x ] I have commented my code, particularly in hard-to-understand areas
~- [ ] I have made corresponding changes to the documentation~
~- [ ] I have made corresponding change to the default configuration files~
~- [ ] I have added tests that prove my fix is effective or that my feature works~
~- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.~

## How to test this PR locally

After running `mage goIntegTest` (or `mage pythonIntegTest`) within a beat, the compose logs should be available under `build/system-tests/docker-logs/TEST-docker-compose-goIntegTest.log`. 

Jenkins will automatically upload the docker-compose service logs to Google Cloud Storage when a test fails. You can manually force Jenkins to upload logs for every test suite by editing appropriate archiving step in the top level Jenkins file.

## Related issues

- Relates [#1821](https://github.com/elastic/observability-dev/issues/1821)
- Relates [#29348](https://github.com/elastic/beats/issues/29348)

## Logs

The libbeat tests produce a single deprecation log (below). I have not found any other logs of interest.

```json
{
  "@timestamp": "2021-12-19T18:13:23.917Z",
  "log.level": "WARN",
  "data_stream.dataset": "deprecation.elasticsearch",
  "data_stream.namespace": "default",
  "data_stream.type": "logs",
  "elasticsearch.event.category": "api",
  "event.code": "deprecated_route_PUT_/_template/{name}",
  "message": "Legacy index templates are deprecated in favor of composable templates.",
  "ecs.version": "1.2.0",
  "service.name": "ES_ECS",
  "event.dataset": "deprecation.elasticsearch",
  "process.thread.name": "elasticsearch[5ca57392301f][transport_worker][T#1]",
  "log.logger": "org.elasticsearch.deprecation.rest.RestController",
  "elasticsearch.cluster.uuid": "mdrgEfcUQWqd2Om68ZQNrQ",
  "elasticsearch.node.id": "gUl4JkToTa-VAL00mPqYSQ",
  "elasticsearch.node.name": "5ca57392301f",
  "elasticsearch.cluster.name": "docker-cluster"
}
```

### CODE REVIEW

**Comment by cmacknz** [comment]
This still needs some work to have Jenkins properly archive the integration test logs it appears.

**Comment by cmacknz** [comment]
I've confirmed this works as expected on Jenkins now. The only thing missing before merge is what branches to target besides master since this isn't a bug fix or feature enhancement.

File: dev-tools/mage/integtest_docker.go
Comment (ruflin) [inline dev-tools/mage/integtest_docker.go:187]: Is the composeLogFile entry needed twice?

**Review by ruflin — APPROVED** [review]
Nice, this will potentially also help debug some other issue. @v1v You likely know the upload part around jenkins best, would be nice if you could have a quick look.

@cmacknz Around the backports, I suggest 8.0 and 7.17. The change only applies to test files so there should not be any risk to get this into 7.17. And we will need these log files especially for 7.17 to check for the deprecation logs. You can use the backport labels to get it automatically backported.

Left one minor question in the review as I stumbled over one of the entries but it might be fine as it all seems to work as expected. Approving so you can get it in.

File: dev-tools/mage/integtest_docker.go
Comment (cmacknz) [reply dev-tools/mage/integtest_docker.go:187]: These two arguments are the stdout and stderr destinations, so if you want to capture everything they are needed. I'll add some inline comments to make this clearer, it does read oddly without them if you don't know what the arguments are.

```go
	_, err = sh.Exec(
		composeEnv,
		composeLogFile, // stdout
		composeLogFile, // stderr
```

File: Jenkinsfile
Comment (v1v) [inline Jenkinsfile:705]: This is my proposal, to store the artifacts `'build/system-tests/docker-logs/TEST-docker-compose-*.log'` stored in each beat.

Comment (v1v) [reply Jenkinsfile:705]: The CI pipeline, execute one stage per beat. Therefore I assumed `build/system-tests/docker-logs/` is a relative path for each beat that executed the system tests.

*For instance*
Given `metricbeat` 
   And guessing it can execute the system-tests
   And the workspace location is `/src/beats`
Then the full path location for those files would be `/src/beats/metricbeat/build/system-tests/docker-logs`

Comment (cmacknz) [reply Jenkinsfile:705]: Yes the docker-logs paths are as you assumed. 

If this will always save and present the log files in the list of build artifacts for each run, then that is much more convenient than the current behaviour of only uploading them to GCS on failures. 

Thanks!

File: dev-tools/mage/integtest_docker.go
Comment (ruflin) [reply dev-tools/mage/integtest_docker.go]: Got it, thx.

**Review by ruflin — APPROVED** [review]

File: Jenkinsfile
Comment (v1v) [reply Jenkinsfile:705]: Thanks Craig, let me put some fix, since I missed the beat folder

image

```suggestion
            archiveArtifacts(allowEmptyArchive: true, artifacts: "${directory}/build/system-tests/docker-logs/TEST-docker-compose-*.log")
```

**Comment by v1v** [comment]
### CI pipeline is now storing the log files


https://beats-ci.elastic.co/job/Beats/job/beats/job/PR-29504/8/artifact/filebeat/build/system-tests/docker-logs/

image

https://beats-ci.elastic.co/job/Beats/job/beats/job/PR-29504/8/artifact/filebeat/build/system-tests/docker-logs/TEST-docker-compose-goIntegTest.log/*view*/

image

**Review by v1v — APPROVED** [review]

**Pull request merged**
